tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp

```diff
@@ -0,0 +1,254 @@
+		phase++;
+		return 0;
+	};
+	target_read_processed_0_fake.custom_fake = [&phase](struct i2c_target_config *,
+							    uint8_t *) -> int {
+		zassert_equal(3, phase, "Expected a call to read_processed as the 4th step");
+		phase++;
+		return 0;
+	};
+	target_stop_0_fake.custom_fake = [&phase](struct i2c_target_config *) -> int {
+		zassert_equal(4, phase, "Expected a call to stop as the 5th step");
+		phase++;
+		return 0;
+	};
+	zassert_ok(i2c_transfer(controller, msgs, ARRAY_SIZE(msgs),
+				emulated_target_config[0].address));
+	zexpect_equal(1, target_write_requested_0_fake.call_count,
+		      "Expected target_write_requested to be called once, but got %d",
+		      target_write_requested_0_fake.call_count);
+	zexpect_equal(1, target_write_received_0_fake.call_count,
+		      "Expected target_write_received to be called once, but got %d",
+		      target_write_received_0_fake.call_count);
+	zexpect_equal(1, target_read_requested_0_fake.call_count,
+		      "Expected target_read_requested to be called once, but got %d",
+		      target_read_requested_0_fake.call_count);
+	zexpect_equal(1, target_read_processed_0_fake.call_count,
+		      "Expected target_read_processed to be called once, but got %d",
+		      target_read_processed_0_fake.call_count);
+	zexpect_equal(1, target_stop_0_fake.call_count,
+		      "Expected target_stop to be called once, but got %d",
+		      target_stop_0_fake.call_count);
+	zexpect_equal(5, phase, "Expected a total of 5 phases, but got %d", phase);
+}
+
+ZTEST(i2c_emul_forwarding, test_forward_two_targets)
+{
+	uint8_t read_data[2];
+
+	// Read the second forward and ensure that we only forwarded to the correct one
+	zassert_ok(i2c_read(controller, read_data, sizeof(read_data),
+			    emulated_target_config[1].address));
+
+	// Check that we got the forward
+	zexpect_equal(1, target_read_requested_1_fake.call_count,
+		      "Expected to be called 1 time, got %d",
+		      target_read_requested_1_fake.call_count);
+	zexpect_equal(1, target_read_processed_1_fake.call_count,
+		      "Expected to be called 1 time, got %d",
+		      target_read_processed_1_fake.call_count);
+
+	// Check that we didn't forward to the first target
+	zexpect_equal(0, target_read_requested_0_fake.call_count,
+		      "Expected to be called 0 times, got %d",
+		      target_read_requested_0_fake.call_count);
+	zexpect_equal(0, target_read_processed_0_fake.call_count,
+		      "Expected to be called 0 times, got %d",
+		      target_read_processed_0_fake.call_count);
+}
+
+ZTEST(i2c_emul_forwarding, test_error_in_write_received)
+{
+	uint8_t data;
+
+	target_write_received_0_fake.return_val = -EINTR;
+	zassert_equal(-EINTR, i2c_write(controller, &data, 1, emulated_target_config[0].address));
+	zexpect_equal(1, target_write_requested_0_fake.call_count);
+	zexpect_equal(1, target_write_received_0_fake.call_count);
+	zexpect_equal(0, target_stop_0_fake.call_count);
+}
+
+} // namespace
```

Comment:
?

Reply:
I'm going to comment in the other thread since these comments are very related